zypp/KeyManager.cc

```diff
@@ -0,0 +1,388 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+#include "zypp/KeyManager.h"
+#include "zypp/PublicKey.h"
+#include "zypp/PathInfo.h"
+#include "zypp/base/Logger.h"
+#include "zypp/TmpPath.h"
+#include "zypp/base/String.h"
+
+#include <boost/thread/once.hpp>
+#include <boost/interprocess/smart_ptr/scoped_ptr.hpp>
+#include <gpgme.h>
+
+#include <stdio.h>
+
+#undef  ZYPP_BASE_LOGGER_LOGGROUP
+#define ZYPP_BASE_LOGGER_LOGGROUP "zypp::gpg"
+
+// @TODO [threading]
+// make sure to call the init code of gpgme only once
+// this might need to be moved to a different location when
+// threads are introduced into libzypp
+boost::once_flag gpgme_init_once = BOOST_ONCE_INIT;
+using std::endl;
+
+//using boost::interprocess pointer because it allows a custom deleter
+typedef boost::interprocess::scoped_ptr<gpgme_data, boost::function<void (gpgme_data_t)>> GpgmeDataPtr;
+typedef boost::interprocess::scoped_ptr<_gpgme_key, boost::function<void (gpgme_key_t)>>  GpgmeKeyPtr;
+typedef boost::interprocess::scoped_ptr<FILE, boost::function<int (FILE *)>> FILEPtr;
+
+struct GpgmeErr
+{
+  GpgmeErr( gpgme_error_t err_r = GPG_ERR_NO_ERROR )
+  : _err( err_r )
+  {}
+  operator gpgme_error_t() const { return _err; }
+  GpgmeErr &operator= (gpgme_error_t err_r) { _err = err_r; return *this;}
```

Comment:
Out of curiosity, did you need to explicitly define the operator= ?

Reply:
Yes, did not compile without it, when I want to assign a new error value to a existing instance the = operator is needed. I guess using GpgmeErr & as argument would've worked too.

Reply:
Which g++ version are you using? 
Just asking because the CTOR allows implicit conversion from gpgme_error_t to GpgmeErr, so trivial assignment should be possible (g++ 4.8.5 and 7.3.0 compiled it without the operator=).